### ISSUE TITLE
[Feature] Community Scripting Functions

### ISSUE THREAD

**Issue by RandoRuda** [BG3-Community-Library-Team/BG3-Community-Library#71]
**Is your feature request related to a problem? Please describe.**
Just having a look at the unpacked Fantastical Races mod and I noticed the scripts & functions they have added. My immediate thought went to the community library and making things easier for newer modders.

**Describe the solution you'd like**
We could look at adding our own functions that we might see lots of modders use. 
I feel like we see a lot of newer mod makers with less coding experience, so potentially we could pre-write a lot of functions that might be needed for the standard D&D-Classes/Races that people might wish to add.
Another example might be writing a simpler function for Pact Weapon casting modifiers.

**Comment by NellsRelo** [comment]
Agreed, I think this is a high priority, as we've got a decent chunk of useable data resources, and many of the existing ones could likely be simplified with custom scripts.

I think defining a few concrete basic scripts via feature requests to pursue may be worthwhile - I've added the Scripts label to help keep these organized.  Many of the ones I see in Fantastic Multiverse seem to be conditionals - I think first, we should look at some of the more complex passives and statuses we have going on, and seeing if we can move some of those into functions. Healing Strike might be a good candidate, from its conditionals - a simple `ActiveWeaponHasStatus(StatusName)` which is just a simpler way of inputting `HasStatus('StatusName', GetActiveWeapon())`, for example. SpellDrinker may also be a good  passive to look at for ideas. Even things in the base game, for example, `HasStatus('BLEEDING',context.Target)` could be reduced to `IsBleeding(context.Target)`

Some function calls in the base game also return a value based on certain parameters, though I'm blanking on specific examples, so they shouldn't all need to be Booleans.

**Comment by NellsRelo** [comment]
I've started pushing a few here: https://github.com/BG3-Community-Library-Team/BG3-Community-Library/pull/78

**Comment by RandoRuda** [comment]
My current hyperfocus has taken me to investigating feats and I noticed there were very few feat conditions (included below).

As I look some of the official material, I see there's a few requirements not already covered:
- Race.
- Level.
- Class.
- Exclusivity (E.g. Eberron Dragonmarks.)

Maybe we could add some to the library to support those either adding some of the other feats or adding feats specific to their class/race mods.

---

```function FeatRequirementProficiency(proficiencyName)
	return HasProficiency(proficiencyName, context.Source)
end```

```function FeatRequirementAbilityGreaterEqual(ability, value)
	if ability == "Strength" then
		return ConditionResult(context.Source.Strength >= value)
	elseif ability == "Dexterity" then
		return ConditionResult(context.Source.Dexterity >= value)
	elseif ability == "Constitution" then
		return ConditionResult(context.Source.Constitution >= value)
	elseif ability == "Intelligence" then
		return ConditionResult(context.Source.Intelligence >= value)
	elseif ability == "Wisdom" then
		return ConditionResult(context.Source.Wisdom >= value)
	elseif ability == "Charisma" then
		return ConditionResult(context.Source.Charisma >= value)
	else
		return ConditionResult(false)
	end
end```

```function FeatRequirementHasAnySpell()
	return ConditionResult(true)
end```

**Comment by Mythalar** [comment]
Hello, I don't know if it's the place to ask but, seing scripting functions are discussed :
Your new functions can be used in a mod without any config if the Community Library is in the mod list?

Asking this since I have the CL (used by some of my mod list), and wanted to use CL_ApplyStatusBulk this way in a new passive (trimmed down to only relevant parts) : 

```
data "StatsFunctorContext" "OnDamage"
data "Conditions" "IsSpell() and not Item() and not Self() and not IsLastConditionRollSuccess(ConditionRollType.ConditionSavingThrow)"
data "StatsFunctors" "IF(HasDamageDoneForType(DamageType.Lightning)):ApplyStatus(SHOCKED, 100, 2);IF(HasDamageDoneForType(DamageType.Cold)):CL_ApplyStatusBulk({'MAG_FROST', 'MAG_FROST_DURATION_TECHNICAL'},{context.Target},{100}, {2, 1})" 
```

But the cold part does not work :/
I did not know those functions were a thing, and they could really be useful for me (so 13 passives could be reduced to only one with CL_ApplyStatusBulk !)

If I have to ask elsewhere feel free to tell me ^^

**Comment by RandoRuda** [comment]
> My current hyperfocus has taken me to investigating feats and I noticed there were very few feat conditions (included below).
> 
> As I look some of the official material, I see there's a few requirements not already covered:
> 
>     * Race.
> 
>     * Level.
> 
>     * Class.
> 
>     * Exclusivity (E.g. Eberron Dragonmarks.)
> 
> 
> Maybe we could add some to the library to support those either adding some of the other feats or adding feats specific to their class/race mods.
> 
> Proficiency
> 
> ```
> function FeatRequirementProficiency(proficiencyName)
> 	return HasProficiency(proficiencyName, context.Source)
> end
> ```
> 
> Ability Score
> 
> ```
> function FeatRequirementAbilityGreaterEqual(ability, value)
> 	if ability == "Strength" then
> 		return ConditionResult(context.Source.Strength >= value)
> 	elseif ability == "Dexterity" then
> 		return ConditionResult(context.Source.Dexterity >= value)
> 	elseif ability == "Constitution" then
> 		return ConditionResult(context.Source.Constitution >= value)
> 	elseif ability == "Intelligence" then
> 		return ConditionResult(context.Source.Intelligence >= value)
> 	elseif ability == "Wisdom" then
> 		return ConditionResult(context.Source.Wisdom >= value)
> 	elseif ability == "Charisma" then
> 		return ConditionResult(context.Source.Charisma >= value)
> 	else
> 		return ConditionResult(false)
> 	end
> end
> ```
> 
> Has Spell
> 
> ```
> function FeatRequirementHasAnySpell()
> 	return ConditionResult(true)
> end
> ```

I did some testing and investigated this recently. Unfortunately, I was unable to get the feats section to recognise any other functions for requirements. :(
